schema/function_test.go

```diff
@@ -0,0 +1,517 @@
+package schema_test
+
+import (
+	"fmt"
+	"go.arcalot.io/assert"
+	"go.flow.arcalot.io/pluginsdk/schema"
+	"reflect"
+	"testing"
+)
+
+// Simple, no input or output.
+func TestCallableFunctionSchema_Simple(t *testing.T) {
+	called := false
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		nil,
+		nil,
+		func() {
+			called = true
+		},
+	)
+	assert.NoError(t, err)
+	assert.Equals(t, called, false)
+	result, err := simpleFunc.Call([]any{})
+	assert.NoError(t, err)
+	assert.Nil(t, result)
+	assert.Equals(t, called, true)
+}
+
+// Simple, with one input.
+func TestCallableFunctionSchema_Simple1Param(t *testing.T) {
+	passedInVal := ""
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)},
+		nil,
+		nil,
+		func(test string) {
+			passedInVal = test
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{"a"})
+	assert.NoError(t, err)
+	assert.Nil(t, result)
+	assert.Equals(t, passedInVal, "a")
+}
+
+// Simple, with no input, and only error out.
+func TestCallableFunctionSchema_SimpleWithNilErr(t *testing.T) {
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		nil,
+		nil,
+		func() error {
+			return nil
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{})
```

Comment:
They are identical. That's just what I ended up using. Sometimes the IDE complains when I use the second format. But I used that format for all of the calls to `Call`.

Reply:
I really doubt that they are _identical_ -- knowing some of the Go philosophy, it seems highly unlikely that they would provide more than one way to do the same thing.  That said, I'm not finding any obvious difference between them in the language specification:  `make([]any, 0)` returns an initialized slice, backed by a "real", hidden, zero-length array, whereas `[]any{}` is a "slice literal", evaluating to an initialized, zero-length slice backed by a "real" (and also "hidden") array (which is presumably also zero-length).  I suspect that the difference is that (barring optimizations), the literal is allocated on the stack or perhaps in the process's static data region (there's lots of room there for zero-length things... 😉), whereas the result of `make()` is allocated from the heap.  (The Go documentation seems to be pretty non-specific about where and how they store things, but the fact that you can take the address of a local variable and it remains valid after leaving the function scope indicates that the implementation can and will allocate things from the heap when needed, even when they otherwise would be allocated on the stack in other languages.)

The fact that your IDE complains about certain usages is a further indication that they are _not_ identical.  However, I couldn't find any guidance on which to use when.  😬  It would be nice if we could follow some sort of consistency....